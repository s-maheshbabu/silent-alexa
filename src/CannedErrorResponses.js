const { Map } = require("immutable");

/**
 * We need to show human readable error messages to users when Alexa responds
 * with error codes or we encounter internal/unknown errors.
 * This is a mapping from Alexa/Custom error codes to canned responses.
 * @link https://developer.amazon.com/docs/alexa-voice-service/exceptions.html
 */
// TODO: Change the place holder error responses to real ones.
const cannedErrorResponses = Map({
  INVALID_REQUEST_EXCEPTION: "a canned response for INVALID_REQUEST_EXCEPTION",
  UNAUTHORIZED_REQUEST_EXCEPTION:
    "a canned response for UNAUTHORIZED_REQUEST_EXCEPTION",
  UNSUPPORTED_MEDIA_TYPE: "a canned response for UNSUPPORTED_MEDIA_TYPE",
  THROTTLING_EXCEPTION: "a canned response for THROTTLING_EXCEPTION",
  INTERNAL_SERVICE_EXCEPTION:
    "a canned response for INTERNAL_SERVICE_EXCEPTION",
  "N/A": "a canned response for N/A Exception",

  UNKNOWN_ERROR: "a canned response for UNKNOWN_ERROR"
});

const customErrorCodes = Object.freeze({
  UNKNOWN_ERROR: "UNKNOWN_ERROR"
});

module.exports = {
  cannedErrorResponses: cannedErrorResponses,
  customErrorCodes: customErrorCodes
};
