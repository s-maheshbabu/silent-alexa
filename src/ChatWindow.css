/** Scafolding **/

.leftpanel {
  display: grid;
  grid-template-rows: auto 30px;
  height: calc(100vh - 60px);
}

.panel-body {
  overflow: auto;
}

/** Styles **/
.leftpanel {
  border-right: 1px solid #d3c5cf;
}

.leftpanel .chat-input {
  display: flex;
  align-items: center;
  justify-content: center;
}
