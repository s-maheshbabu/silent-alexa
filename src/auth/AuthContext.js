import React, { useState } from "react";
import IllegalArgumentError from "errors/IllegalArgumentError";
import Cookies from "js-cookie";
import util from "util";

export const AuthContext = React.createContext();

export const AMAZON_LOGIN_COOKIE = "amazon_Login_accessToken";

export default ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const defaultContext = {
    setLWAResponse: lwaResponse => {
      _persist(lwaResponse);
      setIsAuthenticated(Cookies.get(AMAZON_LOGIN_COOKIE) !== undefined);
    },
    isAuthenticated,
    getAccessToken: () => {
      return Cookies.get(AMAZON_LOGIN_COOKIE);
    },
    clear: () => {
      Cookies.remove(AMAZON_LOGIN_COOKIE);
      setIsAuthenticated(false);
    }
  };
  return (
    <AuthContext.Provider value={defaultContext}>
      {children}
    </AuthContext.Provider>
  );
};

const _persist = lwaResponse => {
  const numberOfSecondsInADay = 86400;

  if (lwaResponse && lwaResponse.access_token && lwaResponse.expires_in) {
    Cookies.set(AMAZON_LOGIN_COOKIE, lwaResponse.access_token, {
      expires: lwaResponse.expires_in / numberOfSecondsInADay,
      secure: false // TODO: Change localhost to also use https and then change this to true.
    });
  } else {
    const serializedLWAResponse = util.inspect(lwaResponse, {
      showHidden: true,
      depth: null
    });
    throw new IllegalArgumentError(
      `LoginWithAmazon Authorization response is undefined or 
        doesnt have access_token/expires_in.
        lwaResponse: ${serializedLWAResponse}`
    );
  }
};
