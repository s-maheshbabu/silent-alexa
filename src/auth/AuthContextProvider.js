import React, { useState } from "react";
import IllegalArgumentError from "errors/IllegalArgumentError";
import Cookies from "js-cookie";
import util from "util";

export const AuthContext = React.createContext();

export const AMAZON_LOGIN_COOKIE = "amazon_Login_accessToken";

// TODO This class is using the presence of AMAZON_LOGIN_COOKIE to mean that the user
// is authenticated. The assumption is that Cookies.get won't return expired cookies. 
// Is it true? If not, we should check for expiration of cookies.
export default ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = useState(Cookies.get(AMAZON_LOGIN_COOKIE) !== undefined);

  const defaultContext = {
    setLWAResponse: lwaResponse => {
      _persist(lwaResponse);
      setIsAuthenticated(Cookies.get(AMAZON_LOGIN_COOKIE) !== undefined);
    },
    isAuthenticated,
    getAccessToken: () => {
      return Cookies.get(AMAZON_LOGIN_COOKIE);
    },
    clear: () => {
      console.log("Auth info cleared========");
      Cookies.remove(AMAZON_LOGIN_COOKIE);
      setIsAuthenticated(false);
    }
  };
  return (
    <AuthContext.Provider value={defaultContext}>
      {children}
    </AuthContext.Provider>
  );
};

const _persist = lwaResponse => {
  const numberOfSecondsInADay = 86400;

  if (lwaResponse && lwaResponse.access_token && lwaResponse.expires_in) {
    Cookies.set(AMAZON_LOGIN_COOKIE, lwaResponse.access_token, {
      expires: lwaResponse.expires_in / numberOfSecondsInADay, // TODO Why are cookies expiring in an hour?
      secure: false // TODO: Change localhost to also use https and then change this to true.
    });
  } else {
    const serializedLWAResponse = util.inspect(lwaResponse, {
      showHidden: true,
      depth: null
    });
    throw new IllegalArgumentError(
      `LoginWithAmazon Authorization response is undefined or 
        doesnt have access_token/expires_in.
        lwaResponse: ${serializedLWAResponse}`
    );
  }
};
