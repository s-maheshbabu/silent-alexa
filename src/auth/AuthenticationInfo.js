import IllegalArgumentError from "errors/IllegalArgumentError";
import Cookies from "js-cookie";
import util from "util";

/**
 * Wrapper class to hold access_token and expires_in returned in LoginWithAmazon
 * authorization response.
 * Instances are created only if access_token and expires_in are defined. The token
 * is stored in a cookie called 'amazon_Login_accessToken' because that is the
 * cookie that LoginWithAmazon will look for.
 * https://developer.amazon.com/docs/login-with-amazon/javascript-sdk-reference.html#setUseCookie
 */
// TODO: Rename this to AuthenticationManager now that the class has been refactored to not
// actually hold the authentication information.
class AuthenticationInfo {
  persist(lwaResponse) {
    const numberOfSecondsInADay = 86400;

    if (lwaResponse && lwaResponse.access_token && lwaResponse.expires_in) {
      Cookies.set(AMAZON_LOGIN_COOKIE, lwaResponse.access_token, {
        expires: lwaResponse.expires_in / numberOfSecondsInADay,
        secure: false // TODO: Change localhost to also use https and then change this to true.
      });
    } else {
      const serializedLWAResponse = util.inspect(lwaResponse, {
        showHidden: true,
        depth: null
      });
      throw new IllegalArgumentError(
        `LoginWithAmazon Authorization response is undefined or 
          doesnt have access_token/expires_in.
          lwaResponse: ${serializedLWAResponse}`
      );
    }
  }

  /**
   * @returns true is authentication info is present and false otherwise.
   * This does not guarantee that the tokens haven't already expired.
   * Callers should be able to handle the case of tokens being
   * invalid when they try to use them.
   */
  isPresent = () => {
    return this.getAccessToken() !== undefined;
  };

  /**
   * @returns access_token to access Alexa API
   */
  getAccessToken = () => {
    return undefined;
  };

  /**
   * @returns deleted the cookie that stores the authentication info
   */
  clear = () => {
    Cookies.remove(AMAZON_LOGIN_COOKIE);
  };
}

// Declare a singleton
const instance = new AuthenticationInfo();
Object.freeze(instance);

export default instance;
export const AMAZON_LOGIN_COOKIE = "amazon_Login_accessToken";
