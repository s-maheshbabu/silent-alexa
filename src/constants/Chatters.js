/*
TODO: Improve Chatters tests. Current tests don't fail if the properties
of a chatter (like their name or avatar) are changed.
*/
import React from "react";
import { Map } from "immutable";
import AlexaRingIcon from "Icons/AlexaRingIcon";

const chatterIds = Object.freeze({
  USER: 0,
  ALEXA: 1
});

const chatters = Map([
  [chatterIds.USER, { name: "You", avatar: undefined }],
  [
    chatterIds.ALEXA,
    {
      avatar: <AlexaRingIcon />
    }
  ]
]);

export { chatterIds, chatters };
