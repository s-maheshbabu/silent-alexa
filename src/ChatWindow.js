import React, { Component } from "react";
import { ChatFeed, Message } from "react-chat-ui";
import ContainerDimensions from "react-container-dimensions";

import ChatBubble from "./ChatBubble";
import { cannedErrorResponses, customErrorCodes } from "./CannedErrorResponses";

import UserRequestToAlexaForm from "./UserRequestToAlexaForm";
import "./ChatWindow.css";

import { users, userIds } from "./ConversingUsers";

import AVSGateway from "./AVSGateway";
const avs = new AVSGateway();

class ChatWindow extends Component {
  constructor() {
    super();
    this.state = {
      messages: [
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        // TODO Adding a bunch of dummy messages to make the ChatFeed overflow window height.
        // This will be useful during the development phase but will be removed once we are confident about the layout.
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        }),
        new Message({
          id: userIds.ALEXA,
          message: "Hey! Alexa here.",
          senderName: users.get(userIds.ALEXA).name
        }),
        new Message({
          id: userIds.YOU,
          message: "Hello Alexa!",
          senderName: users.get(userIds.YOU).name
        })
      ],
      userRequestToAlexa: "",
      is_typing: false
    };

    this.onUserRequestToAlexaSubmit = this.onUserRequestToAlexaSubmit.bind(
      this
    );
    this.handleChangeInUserRequestToAlexa = this.handleChangeInUserRequestToAlexa.bind(
      this
    );
  }

  /*
  Once the user submits the request for Alexa, we need to update the messages
  in the state so the chat window gets updated.
  */
  onUserRequestToAlexaSubmit(event) {
    const userRequestToAlexa = this.state.userRequestToAlexa;
    event.preventDefault();
    if (!userRequestToAlexa || 0 === userRequestToAlexa.length) {
      console.log("Request string for Alexa was empty: " + userRequestToAlexa);
      this.setState({ userRequestToAlexa: "" });
      return false;
    }
    this.pushMessage(userIds.YOU, userRequestToAlexa);
    this.setState({ userRequestToAlexa: "" });

    // TODO: Send the actual access token once we integrate ChatWindow with the
    // state object that contains the real access token. For now, not passing
    // an access_token which means AVS will throw an error.
    avs
      .sendTextMessageEvent(userRequestToAlexa /*, "access_token"*/)
      .then(response => {
        this.pushMessage(userIds.ALEXA, response);
      })
      .catch(error => {
        // TODO: Don't show this as an Alexa bubble. It also doesn't make to show it as a user bubble.
        // Need to find a way to represent this error on the UI.
        this.pushMessage(
          userIds.ALEXA,
          cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR)
        );
      });

    return true;
  }

  pushMessage(userId, message) {
    const user = users.get(userId);
    if (!user) {
      console.log("Unknown userId: " + userId);
      return;
    }
    if (!message || 0 === message.length) {
      console.log("Empty message: " + message);
      return;
    }

    const messagesCopy = this.state.messages.slice(); // for immutability
    const newMessage = new Message({
      id: userId,
      message,
      senderName: user.name
    });
    messagesCopy.push(newMessage);
    this.setState({ messages: messagesCopy });
  }

  /*
  Update the state as the user is typing into the input box
  */
  handleChangeInUserRequestToAlexa(event) {
    event.preventDefault();
    this.setState({ userRequestToAlexa: event.target.value });
  }

  render() {
    return (
      <div id="leftpanel">
        <div className="panel-body">
          <ContainerDimensions>
            {({ height }) => (
              <ChatFeed
                messages={this.state.messages}
                isTyping={this.state.is_typing} // is the recipient typing
                hasInputField={false} // use the default input field that is provided along with ChatFeed component
                showSenderName={false} // show the name of the user who sent the message
                bubblesCentered={false}
                maxHeight={height}
                chatBubble={ChatBubble}
              />
            )}
          </ContainerDimensions>
        </div>
        <div id="chat-input">
          <UserRequestToAlexaForm
            value={this.state.userRequestToAlexa}
            onChange={e => this.handleChangeInUserRequestToAlexa(e)}
            onSubmit={e => this.onUserRequestToAlexaSubmit(e)}
          />
        </div>
      </div>
    );
  }
}

export default ChatWindow;
