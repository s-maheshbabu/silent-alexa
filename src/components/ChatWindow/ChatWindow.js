import React, { Component } from "react";
import { ChatFeed, Message } from "monkas-chat";
import ContainerDimensions from "react-container-dimensions";
import { instanceOf } from 'prop-types';

import ChatBubble from "ChatBubble/ChatBubble";
import { cannedErrorResponses, customErrorCodes } from "CannedErrorResponses";

import UserRequestToAlexaForm from "UserRequestToAlexaForm/UserRequestToAlexaForm";
import "./ChatWindow.css";

import { chatters, chatterIds } from "Chatters";

import AVSGateway from "AVSGateway";
import { withRouter } from 'react-router-dom';
import { withCookies, Cookies } from "react-cookie";

import { AMAZON_LOGIN_COOKIE } from "Constants";
const avs = new AVSGateway();

// TODO: Existing dependency for chat window is not supported. Just making a note of
// a potentially better supported solution https://github.com/FaridSafi/react-native-gifted-chat
class ChatWindow extends Component {
  static propTypes = {
    cookies: instanceOf(Cookies).isRequired
  };

  constructor() {
    super();
    this.state = {
      messages: [
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        // TODO Adding a bunch of dummy messages to make the ChatFeed overflow window height.
        // This will be useful during the development phase but will be removed once we are confident about the layout.
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        }),
        new Message({
          id: chatterIds.ALEXA,
          message: "Hey! Alexa here.",
          avatar: chatters.get(chatterIds.ALEXA).avatar
        }),
        new Message({
          id: chatterIds.USER,
          message: "Hello Alexa!",
          senderName: chatters.get(chatterIds.USER).name
        })
      ],
      userRequestToAlexa: "",
      is_typing: false
    };
  }

  /*
  Once the user submits the request for Alexa, we need to update the messages
  in the state so the chat window gets updated.
  */
  onUserRequestToAlexaSubmit(event) {
    const userRequestToAlexa = this.state.userRequestToAlexa;
    event.preventDefault();
    if (!userRequestToAlexa || 0 === userRequestToAlexa.length) {
      console.log("Request string for Alexa was empty: " + userRequestToAlexa);
      this.setState({ userRequestToAlexa: "" });
      return;
    }

    this.pushMessage(chatterIds.USER, userRequestToAlexa);
    this.setState({ userRequestToAlexa: "" });

    const history = this.props.history;
    const { cookies } = this.props;
    const access_token = cookies.get(AMAZON_LOGIN_COOKIE);
    if (access_token === undefined) {
      cookies.remove(AMAZON_LOGIN_COOKIE, undefined, {
        maxAge: 0,
        secure: false, // TODO: Change localhost to also use https and then change this to true.
        path: "/"
      });

      history.push("/access_denied");
      return;
    }

    avs
      .sendTextMessageEvent(userRequestToAlexa, access_token)
      .then(alexaResponses =>
        alexaResponses.map(alexaResponse =>
          this.pushMessage(chatterIds.ALEXA, alexaResponse)
        )
      )
      .catch(error => {
        // TODO: Don't show this as an Alexa bubble. It also doesn't make sense to show it as a user bubble.
        // Need to find a way to represent this error on the UI.
        this.pushMessage(
          chatterIds.ALEXA,
          cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR)
        );
      });
  }

  pushMessage(userId, message) {
    const user = chatters.get(userId);
    if (!user) {
      console.log("Unknown userId: " + userId);
      return;
    }
    if (!message || 0 === message.length) {
      console.log("Empty message: " + message);
      return;
    }

    const messagesCopy = this.state.messages.slice(); // for immutability
    const newMessage = new Message({
      id: userId,
      message,
      senderName: user.name,
      avatar: user.avatar
    });
    messagesCopy.push(newMessage);
    this.setState({ messages: messagesCopy });
  }

  /*
  Update the state as the user is typing into the input box
  */
  handleChangeInUserRequestToAlexa(event) {
    event.preventDefault();
    this.setState({ userRequestToAlexa: event.target.value });
  }

  render() {
    return (
      <div id="leftpanel">
        <div className="panel-body">
          <ContainerDimensions>
            {({ height }) => (
              <ChatFeed
                messages={this.state.messages}
                isTyping={this.state.is_typing} // is the recipient typing
                hasInputField={false} // use the default input field that is provided along with ChatFeed component
                showSenderName={true} // show the name of the user who sent the message
                bubblesCentered={false}
                maxHeight={height}
                chatBubble={ChatBubble}
              />
            )}
          </ContainerDimensions>
        </div>
        <div id="chat-input">
          <UserRequestToAlexaForm
            value={this.state.userRequestToAlexa}
            onChange={e => this.handleChangeInUserRequestToAlexa(e)}
            onSubmit={e => this.onUserRequestToAlexaSubmit(e)}
          />
        </div>
      </div>
    );
  }
}

// TODO: Use Recomponse to make it more readable. Even better if we can make this class
// a function so we can just use hooks.
export default withRouter(withCookies(ChatWindow));
