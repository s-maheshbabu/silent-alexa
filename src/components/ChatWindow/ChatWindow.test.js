import React from "react";
import { shallow, mount } from "enzyme";
import clone from "clone";
import { List, fromJS } from "immutable";

import { ChatFeed, Message } from "monkas-chat";
import ChatWindow from "./ChatWindow";
import { cannedErrorResponses, customErrorCodes } from "CannedErrorResponses";
import { AuthContext } from "auth/AuthContextProvider";

import {
  mockSendTextMessageEventFunction,
  mockAlexaSuccessResponses
} from "AVSGateway";

import { chatters, chatterIds } from "Chatters";

const CHATFEED_CONTAINER_HEIGHT = 234;
const CHATFEED_CONTAINER_HEIGHT_DEFAULT = 0;

const setHeightElement = function (height) {
  Element.prototype.getBoundingClientRect = jest.fn(() => {
    return {
      height: height
    };
  });
};

jest.mock("AVSGateway");

let chatWindow;
let chatWindowInstance;
let originalState;
let preventDefaultSpy;

beforeEach(() => {
  mockSendTextMessageEventFunction.mockClear();

  preventDefaultSpy = jest.fn();
  chatWindow = shallow(<ChatWindow />);
  chatWindowInstance = chatWindow.instance();
  originalState = clone(chatWindowInstance.state);
});

it("renders correctly without crashing", () => {
  expect(chatWindow).toMatchSnapshot();
});

it("passes height of container to ChatFeed component", () => {
  // Set `height` element.
  setHeightElement(CHATFEED_CONTAINER_HEIGHT);

  const wrapper = mountWithContext(<ChatWindow />, {});
  const chatFeed = wrapper.find(ChatFeed);

  expect(chatFeed.length).toBe(1);
  expect(chatFeed.prop("maxHeight")).toBe(CHATFEED_CONTAINER_HEIGHT);

  wrapper.unmount();

  //Reset `height` element
  setHeightElement(CHATFEED_CONTAINER_HEIGHT_DEFAULT);
});

it("persists a given message in state when pushMessage is called", () => {
  const numberOfMessagesAlreadyInState = originalState.messages.length;
  expect(originalState.messages.length).toBe(numberOfMessagesAlreadyInState);

  const userId = chatterIds.USER;
  const user = chatters.get(userId);
  const message = "test message";
  const expectedMessage = new Message({
    id: userId,
    message,
    senderName: user.name
  });

  chatWindowInstance.pushMessage(userId, message);
  const finalState = chatWindowInstance.state;

  const finalMessages = finalState.messages;
  expect(finalMessages.length).toBe(numberOfMessagesAlreadyInState + 1);
  expect(finalMessages[numberOfMessagesAlreadyInState]).toEqual(
    expectedMessage
  );
});

it("handles gracefully when pushMessage is called with an unknown user", () => {
  const invalidUserid = 1000; // valid values are just 0 and 1

  chatWindowInstance.pushMessage(invalidUserid, "test message");
  const finalState = chatWindowInstance.state;

  // Nothing about the state should have changed.
  expect(finalState).toEqual(originalState);
});

it("handles gracefully when pushMessage is called with an empty or null message", () => {
  const userid = 1;
  const emptyMessage = "";

  chatWindowInstance.pushMessage(userid, emptyMessage);
  let finalState = chatWindowInstance.state;

  // Nothing about the state should have changed.
  expect(finalState).toEqual(originalState);

  let nullMessage;

  chatWindowInstance.pushMessage(userid, nullMessage);
  finalState = chatWindowInstance.state;

  // Nothing about the state should have changed.
  expect(finalState).toEqual(originalState);
});

test("that when a user submits the form, we do not call AVSGateway if the access_token is undefined.", () => {
  const contextValue = {
    isAuthenticated: false
  };
  const chatWindow = mount(<AuthContext.Provider value={contextValue}>
    <ChatWindow />
  </AuthContext.Provider>);
  const chatWindowInstance = chatWindow.instance();

  const userRequestToAlexa = "a dummy user request";
  chatWindowInstance.setState({
    userRequestToAlexa: userRequestToAlexa,
    curr_user: chatters.get(chatterIds.YOU)
  });

  chatWindow
    .find("form")
    .simulate("submit", { preventDefault: preventDefaultSpy });

  expect(mockSendTextMessageEventFunction).not.toHaveBeenCalled();
});

it("handles the user's form submission with request to Alexa and populates the state with the user request and Alexa's response", done => {
  const alexaId = chatterIds.ALEXA;
  const alexa = chatters.get(alexaId);

  const expectedAlexaResponses = mockAlexaSuccessResponses.map(
    alexaResponse =>
      new Message({
        id: alexaId,
        message: alexaResponse,
        senderName: alexa.name,
        avatar: alexa.avatar
      })
  );

  testOnUserRequestToAlexaSubmitHandling(expectedAlexaResponses, done);
});

it("handles the case when AVS throws an error in response to a user request. We should populate the state with the user request and a canned response.", done => {
  // mock the AVSGateway to throw an error.
  mockSendTextMessageEventFunction.mockImplementation(() =>
    Promise.reject(
      new Error(cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR))
    )
  );

  const alexaId = chatterIds.ALEXA;
  const alexa = chatters.get(alexaId);
  const expectedAlexaResponse = List.of(
    new Message({
      id: alexaId,
      message: cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR),
      senderName: alexa.name,
      avatar: alexa.avatar
    })
  );

  testOnUserRequestToAlexaSubmitHandling(expectedAlexaResponse, done);
});

it("handles gracefully when the input form is submitted with a null or empty request string", () => {
  let nullUserRequestToAlexa;
  chatWindowInstance.setState({
    userRequestToAlexa: nullUserRequestToAlexa
  });

  chatWindow
    .find("UserRequestToAlexaForm")
    .simulate("submit", { preventDefault: preventDefaultSpy });

  const finalState = chatWindowInstance.state;

  // Verify that preventDefault() is being called.
  expect(preventDefaultSpy).toHaveBeenCalledTimes(1);

  // Nothing about the state should have changed.
  expect(finalState).toEqual(originalState);
});

it("handles the user's input as they are typing their request (before submission)", () => {
  chatWindowInstance.setState({
    userRequestToAlexa: "some initial value",
    curr_user: 1
  });

  const expectedUserRequestToAlexa = "a dummy user request";
  const event = {
    target: { value: expectedUserRequestToAlexa },
    preventDefault: preventDefaultSpy
  };
  chatWindow.find("UserRequestToAlexaForm").simulate("change", event);

  const finalState = chatWindowInstance.state;
  const finalUserRequestToAlexa = finalState.userRequestToAlexa;

  // Verify that preventDefault() is being called.
  expect(preventDefaultSpy).toHaveBeenCalledTimes(1);
  expect(finalUserRequestToAlexa).toEqual(expectedUserRequestToAlexa);
});

/**
 * Helper method to test the interaction with Alexa. Will simulate a user request and
 * verify that the expected messages are populated into the state.
 * @param {Message} expectedAlexaResponses The expected responses from Alexa to be verified
 * against.
 */
const testOnUserRequestToAlexaSubmitHandling = (
  expectedAlexaResponses,
  done
) => {
  const access_token = "a dummy access token";
  const contextValue = {
    isAuthenticated: true,
    getAccessToken: () => access_token
  };
  const chatWindow = mountWithContext(<ChatWindow />, contextValue);
  const chatWindowInstance = chatWindow.instance();
  const originalState = clone(chatWindowInstance.state);

  const numberOfMessagesAlreadyInState = originalState.messages.length;

  const userRequestToAlexa = "a dummy user request";
  const userId = chatterIds.USER;
  const user = chatters.get(userId);

  chatWindowInstance.setState({
    userRequestToAlexa: userRequestToAlexa,
    curr_user: userId
  });

  chatWindow
    .find("form")
    .simulate("submit", { preventDefault: preventDefaultSpy });

  expect(preventDefaultSpy).toHaveBeenCalledTimes(1);

  expect(mockSendTextMessageEventFunction).toHaveBeenCalledTimes(1);
  expect(mockSendTextMessageEventFunction).toHaveBeenCalledWith(
    userRequestToAlexa,
    access_token
  );

  const expectedUserMessage = new Message({
    id: userId,
    message: userRequestToAlexa,
    senderName: user.name
  });

  setImmediate(() => {
    const finalState = chatWindowInstance.state;
    const finalMessages = finalState.messages;

    // We should have added userMessage and all of Alexa's responses.
    const numberOfExpectedAlexaResponses = expectedAlexaResponses.size;
    const numberOfNewMessagesToGoIntoState = 1 + numberOfExpectedAlexaResponses;

    expect(finalMessages.length).toBe(
      numberOfMessagesAlreadyInState + numberOfNewMessagesToGoIntoState
    );
    expect(finalMessages[numberOfMessagesAlreadyInState]).toEqual(
      expectedUserMessage
    );
    expect(
      fromJS(finalMessages.slice(numberOfMessagesAlreadyInState + 1))
    ).toEqual(expectedAlexaResponses);

    expect(chatWindowInstance.state.userRequestToAlexa).toEqual("");
    done();
  });
};

afterEach(() => {
  chatWindow.unmount();
});

const mountWithContext = (component, contextValue) => {
  return mount(
    <AuthContext.Provider value={contextValue}>
      {component}
    </AuthContext.Provider>)
}
