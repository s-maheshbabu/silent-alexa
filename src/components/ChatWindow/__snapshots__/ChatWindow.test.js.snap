// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without crashing 1`] = `
<div
  id="leftpanel"
>
  <div
    className="panel-body"
  >
    <ContainerDimensions>
      <Component />
    </ContainerDimensions>
  </div>
  <div
    id="chat-input"
  >
    <UserRequestToAlexaForm
      onChange={[Function]}
      onSubmit={[Function]}
      value=""
    />
  </div>
</div>
`;
